### PR TITLE
transform update for generic transforms

### PR DESCRIPTION
**Pull request by tdk-meta**
Looking for Comments, Thanks

### CODE REVIEW

File: src/transform/affine.rs [hunk 1]
```diff
@@ -1,4 +1,69 @@
...
+        Err(e) => AffineTransform {
+            matrix2d_transform: in_array.clone(),
+            matrix2d_transform_inverse: Array2::zeros((2, 2)),
+            inverse_exists: false,
+        },
+    };
+    return transform;
+}
+
+pub struct AffineTransform {
```
Comment (xd009642) [inline src/transform/affine.rs:25]: I think this module is underdocumented so probably worth adding some either module level docs or type docs on the structs and public methods.

**Review by xd009642 — COMMENTED** [review]
Only one comment, re docs, code all looks good :tada:

File: src/transform/affine.rs (hunk 1 above)
Comment (tdk-meta) [reply src/transform/affine.rs:25]: Finally got around to updating this,  added a bunch of comments, let me know what you think.

File: src/transform/affine.rs [hunk 2]
```diff
@@ -1,24 +1,95 @@
+use super::Transform;
 use ndarray::{array, prelude::*};
+use ndarray_linalg::Inverse;
 
+
+pub fn transform_from_2dmatrix(in_array: Array2<f64>) -> AffineTransform {
+    /// converts a matrix into an equivalent Transform
```
Comment (xd009642) [inline src/transform/affine.rs]: Doc comments like this should be above the function signature.

File: src/transform/affine.rs [hunk 3]
```diff
@@ -1,24 +1,95 @@
...
+            matrix2d_transform: in_array.clone(),
+            matrix2d_transform_inverse: Array2::zeros((2, 2)),
+            inverse_exists: false,
+        },
+    };
+    return transform;
+}
+
+pub struct AffineTransform {
+    /// Represents a linear transform of either size 2x2
```
Comment (xd009642) [inline src/transform/affine.rs]: Ditto, I imagine this comment should be above the struct? The idea is these should show up for users when they look at the docs (so run `cargo doc` and have a peruse to see how it looks)

**Comment by tdk-meta** [comment]
@xd009642  happy to hear your feedback on this next round of comment polish.  Also made ComposedTransform public

**Comment by xd009642** [comment]
The doc comments still seem to be below the function signatures instead of above them?

**Comment by tdk-meta** [comment]
Gah.. bad commit, sorry, should be working now.

File: src/transform/mod.rs [hunk 4]
```diff
@@ -1,32 +1,74 @@
...
+pub enum TransformError {
+    InvalidTransform,
+    NonInvertibleTransform, }
+
+impl std::error::Error for TransformError {}
+
+impl Display for TransformError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self { 
+            InvalidTransform => return write!(f, "invalid transform" ),
```
Comment (xd009642) [inline src/transform/mod.rs]: I think here you should do `Self::InvalidTransform` that might fix CI

Comment (tdk-meta) [reply src/transform/mod.rs]: Took TransformerError::InvalidTransform to pass tests... hopefully that works.

**Comment by xd009642** [comment]
Okay just a `cargo fmt` and then we should be good

**Review by xd009642 — APPROVED** [review]
LGTM

**Pull request merged**